### PR TITLE
Fix crash when save folder is empty and manage corrupted savegame file

### PR DESCRIPTION
**Pull request by StraToN**
Fixes https://github.com/godot-escoria/escoria-issues/issues/121
Fixes https://github.com/godot-escoria/escoria-issues/issues/122

### CODE REVIEW

**Review by BHSDuncan — APPROVED** [review]
Personally I wish Godot would split out matches from captures to make them easier to read in code, but oh well.

Without testing the actual branch, this looks good to me.

File: addons/escoria-core/ui_library/menus/load_save/load/load_game.gd [hunk 1]
```diff
@@ -34,12 +34,13 @@ func refresh_savegames():
...
-		if saves_list.has(i):
-			var save_data = saves_list[i]
-			var new_slot = slot_ui_scene.instance()
-			$VBoxContainer/ScrollContainer/slots.add_child(
-				new_slot
-			)
-			new_slot.set_slot_name_date(save_data["name"], save_data["date"])
-			new_slot.connect("pressed", self, "_on_slot_pressed", [i])
+	if not saves_list.empty():
+		for i in range(saves_list.keys().max() + 1):
```
Comment (dploeger) [inline addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: Replace this line with

```gdscript
for save_key in saves_list.keys():
```

then you can delete line 39. Replace "i" with "save_key" then.

File: addons/escoria-core/ui_library/menus/load_save/save/save_game.gd [hunk 2]
```diff
@@ -34,13 +34,14 @@ func refresh_savegames():
...
 	var saves_list = escoria.save_manager.get_saves_list()
-	for i in range(saves_list.keys().max() + 1):
-		if saves_list.has(i):
-			var save_data = saves_list[i]
-			var new_slot = slot_ui_scene.instance()
-			_slots.add_child(new_slot)
-			new_slot.set_slot_name_date(save_data["name"], save_data["date"])
-			new_slot.connect("pressed", self, "_on_slot_pressed", [i])
+	if not saves_list.empty():
+		for i in range(saves_list.keys().max() + 1):
```
Comment (dploeger) [inline addons/escoria-core/ui_library/menus/load_save/save/save_game.gd:38]: See my suggestion in load_game

**Review by dploeger — CHANGES_REQUESTED** [review]
Some optimization remarks.

**Comment by dploeger** [comment]
> Personally I wish Godot would split out matches from captures to make them easier to read in code, but oh well.

wdym? Do you mean, that the capture groups should be easier to address? Godot supports named capture groups. So we could go with `save_(?<save_id>[0-9]{3})\\.tres` and use `match.names["save_id"]`.

File: addons/escoria-core/ui_library/menus/load_save/load/load_game.gd (hunk 1 above)
Comment (BHSDuncan) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: Does he not need `i` further down in the `.connect` call?

Comment (dploeger) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: > Does he not need `i` further down in the `.connect` call?

Yes. He can use `save_key` instead there.

Comment (BHSDuncan) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: Ah, ok. So the keys are numeric, then?

**Comment by BHSDuncan** [comment]
> > Personally I wish Godot would split out matches from captures to make them easier to read in code, but oh well.
> 
> wdym? Do you mean, that the capture groups should be easier to address? Godot supports named capture groups. So we could go with `save_(?<save_id>[0-9]{3})\\.tres` and use `match.names["save_id"]`.

Oh, that's good! In that case, I'll make another suggestion for a change. Thanks!

File: addons/escoria-core/ui_library/menus/load_save/load/load_game.gd (hunk 1 above)
Comment (dploeger) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: I believe so.

File: addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd [hunk 3]
```diff
@@ -70,18 +70,30 @@ func get_saves_list() -> Dictionary:
...
+				)
+			else:
+				var save_game_data = {
+					"date": save_game_res["date"],
+					"name": save_game_res["name"],
+					"game_version": save_game_res["game_version"],
+				}
+
+				var matches = regex.search(nextfile)
+				if matches != null and matches.strings.size() > 1:
```
[the remark this answers is not in the document]
Comment (StraToN) [reply addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd:87]: Good point. I'll change that.

File: addons/escoria-core/ui_library/menus/load_save/load/load_game.gd (hunk 1 above)
Comment (BHSDuncan) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: @StraToN can you confirm? Just want to make sure as I do like @dploeger 's suggestion.

Comment (StraToN) [reply addons/escoria-core/ui_library/menus/load_save/load/load_game.gd:38]: Done

File: addons/escoria-core/ui_library/menus/load_save/save/save_game.gd (hunk 2 above)
Comment (StraToN) [reply addons/escoria-core/ui_library/menus/load_save/save/save_game.gd:38]: Done as well

File: addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd (hunk 3 above)
[the remark this answers is not in the document]
Comment (StraToN) [reply addons/escoria-core/game/core-scripts/save_data/esc_save_manager.gd]: Done

**Review by BHSDuncan — APPROVED** [review]

**Review by dploeger — APPROVED** [review]
Awesome 🎁

**Pull request merged**